trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs

```diff
@@ -0,0 +1,360 @@
+use std::{fmt::Debug, ops::Range, time::Duration};
+
+use tracing::debug;
+
+use crate::versioned::usage_stats::UsageStats;
+
+use super::IdIndexedV1StoreConfig;
+
+/// The configuration parameters used by [PruningStrategy].
+#[derive(Clone, Debug)]
+pub struct PruningConfig {
+    /// The fraction of storage capacity that we aim for when pruning.
+    pub target_capacity_fraction: f64,
+    /// The fraction by which we increase/decrease the `max_pruning_count` when pruning duration is
+    /// outside `optimal_pruning_duration_range`.
+    pub max_pruning_count_change_fraction: f64,
+    /// The range of pruning durations that we consider optimal.
+    pub optimal_pruning_duration_range: Range<Duration>,
+}
+
+impl PruningConfig {
+    /// By default, we aim to prune down to 95% of storage capacity.
+    pub const DEFAULT_TARGET_CAPACITY_FRACTION: f64 = 0.95;
+    /// By default, we increase/decrease `max_pruning_count` by 20%.
+    pub const DEFAULT_CHANGE_FRACTION: f64 = 0.2;
+    /// By default, we consider optimal pruning duration between 0.1 and 0.3 seconds.
+    pub const DEFAULT_OPTIMAL_PRUNING_DURATION_RANGE: Range<Duration> =
+        Duration::from_millis(100)..Duration::from_millis(300);
+
+    pub fn new(
+        target_capacity_fraction: f64,
+        max_pruning_count_change_fraction: f64,
+        optimal_pruning_duration_range: Range<Duration>,
+    ) -> Self {
+        if !(0.0..1.0).contains(&target_capacity_fraction) {
+            panic!(
+                "Invalid pruning strategy parameters: target_capacity_fraction={}",
+                target_capacity_fraction
+            )
+        }
+        if !(0.0..1.0).contains(&max_pruning_count_change_fraction) {
```

Comment:
Similarly, it's not totally obvious to me that 1 is a hard upper limit here. If someone wanted to play with 2x the pruning count change, my intuition isn't to panic to prevent them from playing with it.

_Edit: I see where the number came from now, but stand by the above comment when paired with the suggestion below._

Reply:
I actually wanted increasing the `max_pruning_count` to be harder than decreasing, because downside of being too low is not so bad (we prune less items more frequently) while downside of being too high is that db is blocked for longer.

I added better explanation to the `max_pruning_count_change_fraction` field.

However, if the behavior that you expected is more common way of doing it, I'm willing to change it in the followup PR.